fix(alerts): improve Slack API rate limiting for large workspaces

marcosmamorim opened the pull request:
### SUMMARY

Fixes rate limiting issues when fetching Slack channels in large workspaces (10k+ channels).

**Note:** This PR takes a more comprehensive approach than #35588 by:
- Making retry count configurable instead of hardcoded
- **Preventing** rate limits through throttling rather than just detecting them
- Providing configuration options for different workspace sizes
- Including complete documentation for large workspaces

**Problem:**
Workspaces with 20k+ channels require ~21 paginated API calls. Without throttling, rapid consecutive requests trigger Slack's rate limits ([Tier 2/3: 20-50 req/min](https://api.slack.com/apis/rate-limits)). The previous hardcoded retry count of 2 was insufficient for large workspaces.

**Solution:**
- Add `SLACK_API_RATE_LIMIT_RETRY_COUNT` config (default: 5, previously hardcoded to 2)
- Add detailed progress logging (page count, channels fetched)
- Improve error messages with configuration suggestions
- Update documentation for large workspaces (10k+ channels)

**Test Results (17k+ channel workspace):**

| Total Time | Rate Limits | Success Rate |
|------------|-------------|--------------|
| 67s        | 1/68 pages  | 98.5%        |

**Impact:**
- Backwards compatible: Default values maintain current behavior for small workspaces
- Large workspaces: Users can now configure higher retry counts (5-10) and delays (1.0-2.0s)
- Better observability with detailed logging

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

N/A - Backend-only changes

### TESTING INSTRUCTIONS

**Prerequisites:**
- Set `SLACK_API_TOKEN` in config
- Enable `ALERT_REPORTS` feature flag

**Manual Testing:**
1. Create an Alert/Report with Slack notification
2. Select a Slack channel from the dropdown
3. Verify channels load successfully

**For large workspaces (10k+ channels):**
```python
from datetime import timedelta

# Increase cache timeout to reduce API calls
SLACK_CACHE_TIMEOUT = int(timedelta(days=2).total_seconds())

# Increase retry count if needed
SLACK_API_RATE_LIMIT_RETRY_COUNT = 10
```

Check logs for progress messages: "Fetched page X: Y channels (total: Z)"

### ADDITIONAL INFORMATION

- [x] Has associated issue: Fixes #35661
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

marcosmamorim committed:
fix(alerts): improve Slack API rate limiting for large workspaces

- Add configurable SLACK_API_REQUEST_DELAY (default 0.5s)
- Add configurable SLACK_API_RATE_LIMIT_RETRY_COUNT (default 2)
- Implement throttling between paginated API requests
- Add detailed progress logging for channel fetching
- Improve error messages with configuration suggestions
- Update documentation for large workspaces (10k+ channels)

marcosmamorim committed:
docs: clarify Slack configuration defaults

marcosmamorim committed:
refactor(alerts-reports): remove sleep delays from Slack API

Delays don't improve rate limiting and make it slower.
Use 5 retries instead of 2.

sadpandajoe commented on superset/utils/slack.py:
```diff
@@ -104,7 +133,19 @@ def get_channels_with_search(
+                f"Slack API rate limit exceeded: {ex}. "
+                "For large workspaces (10k+ channels), consider increasing "
+                "SLACK_API_RATE_LIMIT_RETRY_COUNT (current default: 5, try 10) "
+                "and SLACK_CACHE_TIMEOUT (recommended: 7 days). "
```
Same here, if someone has cash higher, then we're suggesting to lower it.

sadpandajoe commented on superset/config.py:
```diff
@@ -1735,6 +1735,11 @@ def EMAIL_HEADER_MUTATOR(  # pylint: disable=invalid-name,unused-argument  # noq
+# Maximum number of retries when Slack API returns rate limit errors
+# Default: 5
+# For workspaces with 10k+ channels, consider increasing to 10
+SLACK_API_RATE_LIMIT_RETRY_COUNT = 5
```
```suggestion
SLACK_API_RATE_LIMIT_RETRY_COUNT = 2
```
Place at 2 to match the existing code behavior. Allow others to update if they need more than 2.

marcosmamorim committed:
fix(alerts-reports): keep retry count at 2 and improve logging

Standardize default to match existing hardcoded value for backwards compatibility.
Reduce log noise and simplify error messages.

marcosmamorim committed:
fix(alerts-reports): handle SlackApiError response type safely

Use getattr() to check status_code since ex.response can be a string.
Fixes failing unit test test_handle_slack_client_error_listing_channels.
